In [4]:
import numpy as np 
import pandas as pd 

In [5]:
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/amp-parkinsons-disease-progression-prediction/train_proteins.csv
/kaggle/input/amp-parkinsons-disease-progression-prediction/train_clinical_data.csv
/kaggle/input/amp-parkinsons-disease-progression-prediction/public_timeseries_testing_util.py
/kaggle/input/amp-parkinsons-disease-progression-prediction/supplemental_clinical_data.csv
/kaggle/input/amp-parkinsons-disease-progression-prediction/train_peptides.csv
/kaggle/input/amp-parkinsons-disease-progression-prediction/amp_pd_peptide/competition.cpython-37m-x86_64-linux-gnu.so
/kaggle/input/amp-parkinsons-disease-progression-prediction/amp_pd_peptide/__init__.py
/kaggle/input/amp-parkinsons-disease-progression-prediction/amp_pd_peptide_310/competition.cpython-310-x86_64-linux-gnu.so
/kaggle/input/amp-parkinsons-disease-progression-prediction/amp_pd_peptide_310/__init__.py
/kaggle/input/amp-parkinsons-disease-progression-prediction/example_test_files/sample_submission.csv
/kaggle/input/amp-parkinsons-disease-progression-pre

# Source datasets

In [6]:
train_proteins = pd.read_csv("/kaggle/input/amp-parkinsons-disease-progression-prediction/train_proteins.csv")
train_peptides = pd.read_csv("/kaggle/input/amp-parkinsons-disease-progression-prediction/train_peptides.csv")
train_clinical = pd.read_csv("/kaggle/input/amp-parkinsons-disease-progression-prediction/train_clinical_data.csv")

In [7]:
train_proteins

visit_id  visit_month  patient_id UniProt       NPX
0            55_0            0          55  O00391   11254.3
1            55_0            0          55  O00533  732430.0
2            55_0            0          55  O00584   39585.8
3            55_0            0          55  O14498   41526.9
4            55_0            0          55  O14773   31238.0
...           ...          ...         ...     ...       ...
232736  58648_108          108       58648  Q9UBX5   27387.8
232737  58648_108          108       58648  Q9UHG2  369437.0
232738  58648_108          108       58648  Q9UKV8  105830.0
232739  58648_108          108       58648  Q9Y646   21257.6
232740  58648_108          108       58648  Q9Y6R7   17953.1

[232741 rows x 5 columns]

In [8]:
train_proteins['NPX'].describe()

count    2.327410e+05
mean     2.712077e+06
std      2.224155e+07
min      8.460820e+01
25%      2.946440e+04
50%      1.135560e+05
75%      5.638940e+05
max      6.138510e+08
Name: NPX, dtype: float64

In [9]:
train_proteins['NPX'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 232741 entries, 0 to 232740
Series name: NPX
Non-Null Count   Dtype  
--------------   -----  
232741 non-null  float64
dtypes: float64(1)
memory usage: 1.8 MB


In [38]:
train_proteins.groupby(['visit_id','UniProt'])['NPX'].mean().reset_index()

visit_id UniProt        NPX
0       10053_0  O00391    9104.27
1       10053_0  O00533  402321.00
2       10053_0  O14773    7150.57
3       10053_0  O14791    2497.84
4       10053_0  O15240   83002.90
...         ...     ...        ...
232736    942_6  Q9UHG2  306621.00
232737    942_6  Q9UKV8   82335.50
232738    942_6  Q9UNU6   24018.70
232739    942_6  Q9Y646   18939.50
232740    942_6  Q9Y6R7   15251.20

[232741 rows x 3 columns]

### Observations: high variance of NPX, 233k of protein data

In [10]:
train_peptides

visit_id  visit_month  patient_id UniProt  \
0            55_0            0          55  O00391   
1            55_0            0          55  O00533   
2            55_0            0          55  O00533   
3            55_0            0          55  O00533   
4            55_0            0          55  O00533   
...           ...          ...         ...     ...   
981829  58648_108          108       58648  Q9UHG2   
981830  58648_108          108       58648  Q9UKV8   
981831  58648_108          108       58648  Q9Y646   
981832  58648_108          108       58648  Q9Y6R7   
981833  58648_108          108       58648  Q9Y6R7   

                                       Peptide  PeptideAbundance  
0                                NEQEQPLGQWHLS          11254.30  
1                                  GNPEPTFSWTK         102060.00  
2                              IEIPSSVQQVPTIIK         174185.00  
3       KPQSAVYSTGSNGILLC(UniMod_4)EAEGEPQPTIK          27278.90  
4                                 SMEQNGPGLEYR          30838.70  
...                                        ...               ...  
981829                         ILAGSADSEGVAAPR         202820.00  
981830                           SGNIPAGTTVDTK         105830.00  
981831                             LALLVDTVGPR          21257.60  
981832        AGC(UniMod_4)VAESTAVC(UniMod_4)R           5127.26  
981833                          GATTSPGVYELSSR          12825.90  

[981834 rows x 6 columns]

In [11]:
train_peptides['PeptideAbundance'].describe()

count    9.818340e+05
mean     6.428902e+05
std      3.377989e+06
min      1.099850e+01
25%      2.817425e+04
50%      7.430830e+04
75%      2.213388e+05
max      1.787520e+08
Name: PeptideAbundance, dtype: float64

In [12]:
train_peptides['PeptideAbundance'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 981834 entries, 0 to 981833
Series name: PeptideAbundance
Non-Null Count   Dtype  
--------------   -----  
981834 non-null  float64
dtypes: float64(1)
memory usage: 7.5 MB


In [39]:
train_peptides.groupby(['visit_id','Peptide'])['PeptideAbundance'].mean().reset_index()

visit_id                                           Peptide  \
0       10053_0                                     AADDTWEPFASGK   
1       10053_0  AAFGQGSGPIMLDEVQC(UniMod_4)TGTEASLADC(UniMod_4)K   
2       10053_0                  AAFTEC(UniMod_4)C(UniMod_4)QAADK   
3       10053_0                                     AAVYHHFISDGVR   
4       10053_0                           ADDKETC(UniMod_4)FAEEGK   
...         ...                                               ...   
981829    942_6                                    YVNKEIQNAVNGVK   
981830    942_6                                        YWGVASFLQK   
981831    942_6                             YYC(UniMod_4)FQGNQFLR   
981832    942_6                                         YYTYLIMNK   
981833    942_6                                     YYWGGQYTWDMAK   

        PeptideAbundance  
0              6580710.0  
1                31204.4  
2              7735070.0  
3                46620.3  
4               236144.0  
...                  ...  
981829           80001.2  
981830           79661.9  
981831          573300.0  
981832           48005.8  
981833           15674.1  

[981834 rows x 3 columns]

### Observations: high variance of Abundance, 982k of peptide data

In [13]:
train_clinical

visit_id  patient_id  visit_month  updrs_1  updrs_2  updrs_3  updrs_4  \
0         55_0          55            0     10.0      6.0     15.0      NaN   
1         55_3          55            3     10.0      7.0     25.0      NaN   
2         55_6          55            6      8.0     10.0     34.0      NaN   
3         55_9          55            9      8.0      9.0     30.0      0.0   
4        55_12          55           12     10.0     10.0     41.0      0.0   
...        ...         ...          ...      ...      ...      ...      ...   
2610  65043_48       65043           48      7.0      6.0     13.0      0.0   
2611  65043_54       65043           54      4.0      8.0     11.0      1.0   
2612  65043_60       65043           60      6.0      6.0     16.0      1.0   
2613  65043_72       65043           72      3.0      9.0     14.0      1.0   
2614  65043_84       65043           84      7.0      9.0     20.0      3.0   

     upd23b_clinical_state_on_medication  
0                                    NaN  
1                                    NaN  
2                                    NaN  
3                                     On  
4                                     On  
...                                  ...  
2610                                 Off  
2611                                 Off  
2612                                 Off  
2613                                 Off  
2614                                 Off  

[2615 rows x 8 columns]

In [14]:
train_clinical.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2615 entries, 0 to 2614
Data columns (total 8 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   visit_id                             2615 non-null   object 
 1   patient_id                           2615 non-null   int64  
 2   visit_month                          2615 non-null   int64  
 3   updrs_1                              2614 non-null   float64
 4   updrs_2                              2613 non-null   float64
 5   updrs_3                              2590 non-null   float64
 6   updrs_4                              1577 non-null   float64
 7   upd23b_clinical_state_on_medication  1288 non-null   object 
dtypes: float64(4), int64(2), object(2)
memory usage: 163.6+ KB


In [40]:
len(pd.unique(train_clinical['visit_id']))

2615

In [15]:
train_clinical['visit_month'].value_counts()

0      248
12     243
24     243
36     226
48     196
6      192
18     187
30     173
60     166
42     154
3      115
54     110
84     100
9       99
72      93
96      58
108     12
Name: visit_month, dtype: int64

In [16]:
train_clinical['updrs_1'].value_counts()

5.0     227
1.0     221
2.0     218
4.0     211
6.0     209
3.0     188
7.0     171
8.0     167
9.0     147
0.0     145
10.0    114
11.0     99
12.0     87
15.0     65
14.0     61
13.0     55
16.0     48
17.0     41
20.0     29
18.0     29
19.0     20
21.0     14
23.0     10
25.0      8
26.0      8
22.0      7
28.0      5
27.0      4
31.0      3
24.0      1
29.0      1
33.0      1
Name: updrs_1, dtype: int64

In [17]:
train_clinical['updrs_2'].value_counts()

0.0     501
3.0     177
2.0     173
6.0     168
4.0     163
5.0     160
1.0     155
8.0     146
7.0     125
9.0     115
10.0    104
11.0     87
13.0     79
15.0     67
14.0     61
12.0     60
16.0     50
18.0     44
19.0     32
17.0     27
20.0     27
21.0     16
22.0     16
23.0     15
24.0     11
25.0      8
26.0      7
28.0      6
27.0      6
30.0      1
31.0      1
35.0      1
40.0      1
29.0      1
39.0      1
36.0      1
Name: updrs_2, dtype: int64

In [18]:
train_clinical['updrs_3'].value_counts()

0.0     324
1.0     104
2.0      81
19.0     80
21.0     76
       ... 
78.0      1
64.0      1
85.0      1
86.0      1
65.0      1
Name: updrs_3, Length: 72, dtype: int64

In [19]:
train_clinical['updrs_4'].value_counts()

0.0     961
3.0     107
4.0      86
2.0      83
5.0      82
1.0      64
6.0      55
7.0      40
8.0      27
9.0      22
10.0     17
11.0      8
13.0      7
14.0      7
12.0      5
16.0      3
20.0      1
15.0      1
17.0      1
Name: updrs_4, dtype: int64

In [20]:
train_clinical['upd23b_clinical_state_on_medication'].value_counts()

On     775
Off    513
Name: upd23b_clinical_state_on_medication, dtype: int64

In [21]:
train_clinical['on_medication'] = train_clinical['upd23b_clinical_state_on_medication'].apply(lambda x: 1 if x=='On' else 0)
train_clinical['on_medication'].value_counts()

0    1840
1     775
Name: on_medication, dtype: int64

### Observations: 2615 unique visit_id, 4 labels: updrs (some of them are nulls), 2 features: visit_month and on_medication (no nulls)

# Data preparation

In [46]:
import warnings
warnings.filterwarnings("ignore")

In [47]:
from sklearn.model_selection import train_test_split
import lightgbm as lgb

### note: LightGBM chosen for regression in many trials

In [48]:
def smape(A, F):
    return 100/len(A) * np.sum(2 * np.abs(F - A) / (np.abs(A) + np.abs(F)))

### sMAPE: Metric for the project, it returns values between 0 and 200

In [41]:
def prepare_dataset(train_clinical, train_proteins, train_peptides):
        
    df_protein_grouped = train_proteins.groupby(['visit_id','UniProt'])['NPX'].mean().reset_index()
    df_peptide_grouped = train_peptides.groupby(['visit_id','Peptide'])['PeptideAbundance'].mean().reset_index()
    
    df_protein = df_protein_grouped.pivot(index='visit_id',columns = 'UniProt', values = 'NPX').rename_axis(columns=None).reset_index()
    df_peptide = df_peptide_grouped.pivot(index='visit_id',columns = 'Peptide', values = 'PeptideAbundance').rename_axis(columns=None).reset_index()
    
    df = train_clinical[['visit_id', 'updrs_1', 'updrs_2', 'updrs_3', 'updrs_4', 'visit_month', 'on_medication']]
    df = df.merge(df_protein, on=['visit_id'], how='left')
    df = df.merge(df_peptide, on=['visit_id'], how='left')
    
    return df

### * features: names of proteins and peptides, 
### * values: average NPX for each protein and average Abundance for each peptide, 
### * data assigned to each visit_id

In [74]:
df = prepare_dataset(train_clinical, train_proteins, train_peptides)   
df   # joined data

visit_id  updrs_1  updrs_2  updrs_3  updrs_4  visit_month  \
0         55_0     10.0      6.0     15.0      NaN            0   
1         55_3     10.0      7.0     25.0      NaN            3   
2         55_6      8.0     10.0     34.0      NaN            6   
3         55_9      8.0      9.0     30.0      0.0            9   
4        55_12     10.0     10.0     41.0      0.0           12   
...        ...      ...      ...      ...      ...          ...   
2610  65043_48      7.0      6.0     13.0      0.0           48   
2611  65043_54      4.0      8.0     11.0      1.0           54   
2612  65043_60      6.0      6.0     16.0      1.0           60   
2613  65043_72      3.0      9.0     14.0      1.0           72   
2614  65043_84      7.0      9.0     20.0      3.0           84   

      on_medication   O00391    O00533   O00584  ...  YSLTYIYTGLSK  YTTEIIK  \
0                 0  11254.3  732430.0  39585.8  ...      201158.0  16492.3   
1                 0      NaN       NaN      NaN  ...           NaN      NaN   
2                 0  13163.6  630465.0  35220.8  ...      171079.0  13198.8   
3                 1      NaN       NaN      NaN  ...           NaN      NaN   
4                 1  15257.6  815083.0  41650.9  ...      231772.0  17873.8   
...             ...      ...       ...      ...  ...           ...      ...   
2610              0  10589.6  902434.0  44890.8  ...      233567.0  14478.3   
2611              0      NaN       NaN      NaN  ...           NaN      NaN   
2612              0      NaN       NaN      NaN  ...           NaN      NaN   
2613              0      NaN       NaN      NaN  ...           NaN      NaN   
2614              0      NaN       NaN      NaN  ...           NaN      NaN   

      YVGGQEHFAHLLILR  YVM(UniMod_35)LPVADQDQC(UniMod_4)IR  \
0           3810270.0                             106894.0   
1                 NaN                                  NaN   
2           4119520.0                             113385.0   
3                 NaN                                  NaN   
4           5474140.0                             116286.0   
...               ...                                  ...   
2610        3185530.0                              48793.0   
2611              NaN                                  NaN   
2612              NaN                                  NaN   
2613              NaN                                  NaN   
2614              NaN                                  NaN   

      YVMLPVADQDQC(UniMod_4)IR  YVNKEIQNAVNGVK  YWGVASFLQK  \
0                     580667.0        131155.0    165851.0   
1                          NaN             NaN         NaN   
2                     514861.0        103512.0    144607.0   
3                          NaN             NaN         NaN   
4                     711815.0        136943.0    181763.0   
...                        ...             ...         ...   
2610                  501159.0        133992.0    170146.0   
2611                       NaN             NaN         NaN   
2612                       NaN             NaN         NaN   
2613                       NaN             NaN         NaN   
2614                       NaN             NaN         NaN   

      YYC(UniMod_4)FQGNQFLR  YYTYLIMNK  YYWGGQYTWDMAK  
0                  437305.0    46289.2        14898.4  
1                       NaN        NaN            NaN  
2                  457891.0    40047.7        20703.9  
3                       NaN        NaN            NaN  
4                  452253.0    54725.1        21841.1  
...                     ...        ...            ...  
2610               359045.0    45780.0        17370.6  
2611                    NaN        NaN            NaN  
2612                    NaN        NaN            NaN  
2613                    NaN        NaN            NaN  
2614                    NaN        NaN            NaN  

[2615 rows x 1202 columns]

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2615 entries, 0 to 2614
Columns: 1202 entries, visit_id to YYWGGQYTWDMAK
dtypes: float64(1199), int64(2), object(1)
memory usage: 24.0+ MB


In [83]:
df.isna().sum().sum()

1960979

### note: many nulls

In [93]:
def null_to_mean(dataset, col):
    df = dataset.copy()
    means = {}
    months = pd.unique(df['visit_month'])
    for month in months:
        means[month] = df[df['visit_month']==month][col].mean()
        if pd.isna(means[month]):
            means[month]=0
            
    df[col] = df.apply(lambda x: means[x.visit_month] if pd.isna(x[col]) else x[col], axis = 1)
    return df

In [94]:
def feature_engineering(dataset):

    df = dataset.copy()
    
    df.dropna(thresh = 610, inplace = True)   # min about half of proteines or peptides
    
    null_features = df.columns[7:]            # leave nulls in updrs for now
    for feat in null_features:
        df = null_to_mean(df, feat)
            
    return df

In [95]:
df_eng = feature_engineering(df)   # nulls replacement by means for the same 'visit_month'
df_eng

visit_id  updrs_1  updrs_2  updrs_3  updrs_4  visit_month  \
0         55_0     10.0      6.0     15.0      NaN            0   
2         55_6      8.0     10.0     34.0      NaN            6   
4        55_12     10.0     10.0     41.0      0.0           12   
8        55_36     17.0     18.0     51.0      0.0           36   
15       942_6      8.0      2.0     21.0      NaN            6   
...        ...      ...      ...      ...      ...          ...   
2598  64674_84     11.0     15.0     45.0      4.0           84   
2600   65043_0      2.0      6.0     16.0      NaN            0   
2604  65043_12      4.0      7.0     14.0      0.0           12   
2606  65043_24      4.0      8.0      NaN      0.0           24   
2610  65043_48      7.0      6.0     13.0      0.0           48   

      on_medication     O00391     O00533   O00584  ...  YSLTYIYTGLSK  \
0                 0  11254.300   732430.0  39585.8  ...      201158.0   
2                 0  13163.600   630465.0  35220.8  ...      171079.0   
4                 1  15257.600   815083.0  41650.9  ...      231772.0   
8                 1  13530.800   753832.0  43048.9  ...      185290.0   
15                0  11218.700   399518.0  20581.0  ...      226314.0   
...             ...        ...        ...      ...  ...           ...   
2598              0  11550.134   190487.0  24907.9  ...      203523.0   
2600              0  13472.400   927954.0  42661.5  ...      257361.0   
2604              0  14134.900   984651.0  28990.8  ...      230437.0   
2606              0  14659.500  1062020.0  46440.4  ...      251228.0   
2610              0  10589.600   902434.0  44890.8  ...      233567.0   

       YTTEIIK  YVGGQEHFAHLLILR  YVM(UniMod_35)LPVADQDQC(UniMod_4)IR  \
0     16492.30     3.810270e+06                             106894.0   
2     13198.80     4.119520e+06                             113385.0   
4     17873.80     5.474140e+06                             116286.0   
8     18580.50     2.659660e+06                              90936.9   
15     6399.80     3.469825e+06                              57571.4   
...        ...              ...                                  ...   
2598   3835.58     4.901220e+06                              40325.9   
2600  18316.60     2.514660e+06                              51444.6   
2604  16703.20     2.481560e+06                              44405.0   
2606  18326.20     2.939460e+06                              50588.2   
2610  14478.30     3.185530e+06                              48793.0   

      YVMLPVADQDQC(UniMod_4)IR  YVNKEIQNAVNGVK  YWGVASFLQK  \
0                     580667.0        131155.0    165851.0   
2                     514861.0        103512.0    144607.0   
4                     711815.0        136943.0    181763.0   
8                     679163.0        128593.0    203680.0   
15                    480951.0         80001.2     79661.9   
...                        ...             ...         ...   
2598                  335625.0         49250.4     64076.3   
2600                  530245.0        156148.0    157548.0   
2604                  543391.0        159828.0    161207.0   
2606                  597869.0        148032.0    192857.0   
2610                  501159.0        133992.0    170146.0   

      YYC(UniMod_4)FQGNQFLR  YYTYLIMNK  YYWGGQYTWDMAK  
0                  437305.0    46289.2        14898.4  
2                  457891.0    40047.7        20703.9  
4                  452253.0    54725.1        21841.1  
8                  498621.0    52792.7        13973.7  
15                 573300.0    48005.8        15674.1  
...                     ...        ...            ...  
2598               667993.0    38472.5        21949.1  
2600               336625.0    48423.2        10915.8  
2604               330337.0    45368.1        19023.2  
2606               388125.0    65101.0        20790.1  
2610               359045.0    45780.0        17370.6  

[1067 rows x 1202 columns]

In [96]:
df_eng.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1067 entries, 0 to 2610
Columns: 1202 entries, visit_id to YYWGGQYTWDMAK
dtypes: float64(1199), int64(2), object(1)
memory usage: 9.8+ MB


In [98]:
df_eng.iloc[:,6:].isna().sum().sum()

0

### * rows with more than half features as nulls removed from dataset, 
### * then nulls in columns replaced with mean value for entries with the same visit_month, 
### * no nulls in result

In [60]:
def divide_dataset(df):   # drop nulls of y in divided datasets
    
    df1 = df.copy()
    df1.dropna(subset=['updrs_1'], inplace = True)   
    df1.drop(columns=['visit_id','updrs_2', 'updrs_3', 'updrs_4'], inplace = True)
    
    df2 = df.copy()
    df2.dropna(subset=['updrs_2'], inplace = True)
    df2.drop(columns=['visit_id','updrs_1', 'updrs_3', 'updrs_4'], inplace = True)
    
    df3 = df.copy()
    df3.dropna(subset=['updrs_3'], inplace = True)
    df3.drop(columns=['visit_id','updrs_1', 'updrs_2', 'updrs_4'], inplace = True)
    
    df4 = df.copy()
    df4.dropna(subset=['updrs_4'], inplace = True)
    df4.drop(columns=['visit_id','updrs_1', 'updrs_2', 'updrs_3'], inplace = True)
    
    return df1, df2, df3, df4

### note: 4 datasets for 4 labels (labels always in column 0)

# Predictions

In [51]:
def get_training_subset(df):
    
    X = df.iloc[:, 1:]
    y = df.iloc[:, 0]
    
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size = 0.2, random_state=42)
    
    return X_train, X_val, y_train, y_val

In [52]:
def get_training_subsets(df1, df2, df3, df4):
        
    dataset = []
    
    data = get_training_subset(df1)
    dataset.append((data))
    
    data = get_training_subset(df2)
    dataset.append((data))
    
    data = get_training_subset(df3)
    dataset.append((data))
    
    data = get_training_subset(df4)
    dataset.append((data))

    return dataset

In [54]:
def predict_label(Xt, Xv, yt, yv):
    lgbm_model = lgb.LGBMRegressor(metric = 'mse', early_stopping_round = 10, random_state = 42, verbose= -100)    
    lgbm_model.fit(Xt, yt, eval_set = (Xv, yv), verbose= False)
    yp = lgbm_model.predict(Xv)
    smape_score = smape(yv, yp)
    return smape_score

In [55]:
def predict_labels(data):
    
    X1_train, X1_val, y1_train, y1_val = data[0]
    X2_train, X2_val, y2_train, y2_val = data[1]
    X3_train, X3_val, y3_train, y3_val = data[2]
    X4_train, X4_val, y4_train, y4_val = data[3]
    
    smape_score = []
    
    print("\nPREDICTION OF LABELS")
    print("=======================================")
    
    smape = predict_label(X1_train, X1_val, y1_train, y1_val)
    smape_score.append(smape)
    
    smape = predict_label(X2_train, X2_val, y2_train, y2_val)
    smape_score.append(smape)
    
    smape = predict_label(X3_train, X3_val, y3_train, y3_val)
    smape_score.append(smape)
    
    smape = predict_label(X4_train, X4_val, y4_train, y4_val)
    smape_score.append(smape)

    for ind, smape in enumerate(smape_score):
        print(f"sMAPE for label {ind}: {smape}")
    
    y_lengths = np.array([len(data[0][2]), len(data[1][2]), len(data[2][2]), len(data[3][2])])
    weights = y_lengths / sum(y_lengths)
    smape_score_wav = sum(weights * np.array(smape_score))
    
    print(f"\nsMAPE weighted average of labels: {smape_score_wav:.4f}")

In [99]:
df1, df2, df3, df4 = divide_dataset(df_eng)         # separate datasets for labels with features
subsets = get_training_subsets(df1, df2, df3, df4)  # train and validation for each label
predict_labels(subsets)                             # result of feature engineering


PREDICTION OF LABELS
sMAPE for label 0: 61.475720528341824
sMAPE for label 1: 98.08914834094847
sMAPE for label 2: 85.37979693096374
sMAPE for label 3: 150.61364745273943

sMAPE weighted average of labels: 92.0577


### let's check out if another nulls replacement were better: just zeros instead of means

In [100]:
def feature_engineering_2(dataset):
    
    df = dataset.copy()
    
    df.dropna(thresh = 610, inplace = True)   # min about half of proteines or peptides
    
    null_features = df.columns[7:]            # leave nulls in updrs for now
    for feat in null_features:
        df[feat].fillna(0, inplace = True)
    
    return df

In [101]:
df_eng_2 = feature_engineering_2(df)   # nulls replacement by 0

In [133]:
df1, df2, df3, df4 = divide_dataset(df_eng_2)       # separate datasets for labels with features
subsets = get_training_subsets(df1, df2, df3, df4)  # train and validation for each label
predict_labels(subsets)                             # result of feature engineering


PREDICTION OF LABELS
sMAPE for label 0: 60.54809744181341
sMAPE for label 1: 95.82379536060667
sMAPE for label 2: 83.97191704131686
sMAPE for label 3: 149.0455723512696

sMAPE weighted average of labels: 90.5185


### note: it's a little bit better! (90.52 vs. 92.06)

# Clustering data

In [106]:
from sklearn.cluster import KMeans

In [132]:
def predict_clustered_label(df, clust_nr, verbose = True):
    
    model_clust_km = KMeans(n_clusters = clust_nr, init='k-means++', n_init='auto', random_state=42)
    y_clust = model_clust_km.fit_predict(df.iloc[:,1:])
    
    df_clust = df.copy()
    df_clust['cluster'] = y_clust   # for validation calculated not predicted yet
    
    cl_numbers = y_clust.tolist()
    clusters = np.unique(cl_numbers)
    
    cl_entries = []
    smape_score = []
    
    for cluster in clusters:
    
        df_cl = df_clust[df_clust['cluster'] == cluster].drop(columns=['cluster'])
        
        if len(df_cl)>=8:
            X_train, X_val, y_train, y_val = get_training_subset(df_cl)
            smape = predict_label(X_train, X_val, y_train, y_val)
        
            entries = cl_numbers.count(cluster)
            cl_entries.append(entries)
            smape_score.append(smape)
        
        else:
            cl_entries.append(0)
            smape_score.append(0)
    
    weights = np.array(cl_entries) / sum(cl_entries)
    smape_score_wav = sum(weights * np.array(smape_score))
    
    if verbose:
        for ind, cluster in enumerate(clusters):
            print(f"CLUSTER {cluster}:sMAPE: {smape_score[ind]:.4f}, entries: {cl_entries[ind]}")
    
    print(f"sMAPE weighted average in clusters: {smape_score_wav:.4f}, entries: {sum(cl_entries)}")
    
    return smape_score_wav

In [104]:
def predict_clustered_labels(df1, df2, df3, df4, clust_nr_list):
    
    dfs = (df1.copy(), df2.copy(), df3.copy(), df4.copy())
    smape_score = []
    
    print("\nPREDICTION OF CLUSTERED LABELS")
    print("=================================================")
    
    for ind, df in enumerate(dfs):
        print(f"\nLABEL: {df.columns[0]}")
        smape = predict_clustered_label(df, clust_nr_list[ind])
        smape_score.append(smape)
        print("------------------------------------------------")
            
    y_lengths = np.array([len(dfs[0]), len(dfs[1]), len(dfs[2]), len(dfs[3])])
    weights = y_lengths / sum(y_lengths)
    smape_score_wav = sum(weights * np.array(smape_score))
    
    print(f"\nsMAPE weighted average of labels: {smape_score_wav:.4f}")

### functions above divide features for each label into given number of clusters, and calculate sMAPEs for predictions in each cluster separately

In [134]:
predict_labels(subsets)                                        # 4 predictions
predict_clustered_labels(df1, df2, df3, df4, [3, 3, 3, 3])     # random number of clusters yet


PREDICTION OF LABELS
sMAPE for label 0: 60.54809744181341
sMAPE for label 1: 95.82379536060667
sMAPE for label 2: 83.97191704131686
sMAPE for label 3: 149.0455723512696

sMAPE weighted average of labels: 90.5185

PREDICTION OF CLUSTERED LABELS

LABEL: updrs_1
CLUSTER 0:sMAPE: 60.9966, entries: 345
CLUSTER 1:sMAPE: 79.1236, entries: 183
CLUSTER 2:sMAPE: 66.9839, entries: 539
sMAPE weighted average in clusters: 67.1300, entries: 1067
------------------------------------------------

LABEL: updrs_2
CLUSTER 0:sMAPE: 87.7607, entries: 345
CLUSTER 1:sMAPE: 99.3689, entries: 183
CLUSTER 2:sMAPE: 95.1693, entries: 539
sMAPE weighted average in clusters: 93.4941, entries: 1067
------------------------------------------------

LABEL: updrs_3
CLUSTER 0:sMAPE: 80.0867, entries: 189
CLUSTER 1:sMAPE: 88.3048, entries: 336
CLUSTER 2:sMAPE: 91.6579, entries: 532
sMAPE weighted average in clusters: 88.5230, entries: 1057
------------------------------------------------

LABEL: updrs_4
CLUSTER 0:sMAPE:

### note: for updrs 2 i 4 clustering with random cluster numbers improved sMAPE (93.49 vs. 95.82 and 144.06 vs. 149.04)

# Reduction of features number

In [114]:
def sort_label_correlations(df):   
    
    features = df.columns[1:]
    label = df.columns[0]

    corr_y = {}
    for feat in features:
        corr_y[feat] = abs(df[label].corr(df[feat]))
    
    corr_sorted = dict(sorted(corr_y.items(), reverse=True, key=lambda item: item[1]))
    
    return corr_sorted

### function above returns features sorted by correlation with label in reversed order

In [115]:
def search_features_for_label(dataset, feat_step, max_clust):
    
    sorted_correlations = sort_label_correlations(dataset)
    sorted_features = list(sorted_correlations.keys())
    label = [dataset.columns[0]]
    
    min_smape_feat = 0
    min_smape_clus = 0
    min_smape = 200
    
    for feat_ind in range(feat_step, len(sorted_features), feat_step):
        cols = label + sorted_features[:feat_ind]
        df = dataset[cols]
        
        Xt, Xv, yt, yv = get_training_subset(df)
        smape_lab = predict_label(Xt, Xv, yt, yv)
        
        print(f"\nFeatures number: {feat_ind}, sMAPE: {smape_lab}")
        if smape_lab < min_smape:
            min_smape = smape_lab
            min_smape_feat = feat_ind
            min_smape_clus = 1
        
        clust_numbers = np.arange(2, max_clust+1)
        for clust_nr in clust_numbers:
            if clust_nr < feat_ind/20:
                print(f"number of clusters: {clust_nr}")
                smape_clust = predict_clustered_label(df, clust_nr, verbose = False)
                
                if smape_clust < min_smape:
                    min_smape = smape_clust
                    min_smape_feat = feat_ind
                    min_smape_clus = clust_nr
        
    return min_smape, min_smape_feat, min_smape_clus

In [125]:
def search_features_for_labels(df1, df2, df3, df4, feat_step, max_clust):
    
    label1 = df1.columns[0]
    label2 = df2.columns[0]
    label3 = df3.columns[0]
    label4 = df4.columns[0]
    
    print(f"\nSearching parameters for label {label1}")
    min_smape1, min_smape_feat1, min_smape_clus1 = search_features_for_label(df1, feat_step, max_clust)
    
    print("---------------------------------------")
    print(f"\nSearching parameters for label {label2}")
    min_smape2, min_smape_feat2, min_smape_clus2 = search_features_for_label(df2, feat_step, max_clust)
    
    print("---------------------------------------")
    print(f"\nSearching parameters for label {label3}")
    min_smape3, min_smape_feat3, min_smape_clus3 = search_features_for_label(df3, feat_step, max_clust)
    
    print("---------------------------------------")
    print(f"\nSearching parameters for label {label4}")
    min_smape4, min_smape_feat4, min_smape_clus4 = search_features_for_label(df4, feat_step, max_clust)
    
    print("\n\nSUMMARY")
    print("==================================================")
    print(f"\nLABEL {label1}")
    print(f"Min sMAPE: {min_smape1} for features number: {min_smape_feat1} and clusters number: {min_smape_clus1}")
    print(f"\nLABEL {label2}")
    print(f"Min sMAPE: {min_smape2} for features number: {min_smape_feat2} and clusters number: {min_smape_clus2}")
    print(f"\nLABEL {label3}")
    print(f"Min sMAPE: {min_smape3} for features number: {min_smape_feat3} and clusters number: {min_smape_clus3}")
    print(f"\nLABEL {label4}")
    print(f"Min sMAPE: {min_smape4} for features number: {min_smape_feat4} and clusters number: {min_smape_clus4}")
    
    y_lengths = np.array([len(df1), len(df2), len(df3), len(df4)])
    weights = y_lengths / sum(y_lengths)
    smape_score_wav = sum(weights * np.array([min_smape1, min_smape2, min_smape3, min_smape4]))
    
    print(f"\nsMAPE weighted average of labels: {smape_score_wav:.4f}")

### functions above calculate sMAPEs for combination of different numbers of features (beginning from most correlated) and different number of clusters

In [135]:
search_features_for_labels(df1, df2, df3, df4, 300, 2)


Searching parameters for label updrs_1

Features number: 300, sMAPE: 62.86822130541398
number of clusters: 2
sMAPE weighted average in clusters: 63.9158, entries: 1067

Features number: 600, sMAPE: 63.1810663427057
number of clusters: 2
sMAPE weighted average in clusters: 65.6077, entries: 1067

Features number: 900, sMAPE: 63.82325960786169
number of clusters: 2
sMAPE weighted average in clusters: 59.6234, entries: 1067
---------------------------------------

Searching parameters for label updrs_2

Features number: 300, sMAPE: 99.11359798430748
number of clusters: 2
sMAPE weighted average in clusters: 91.9457, entries: 1067

Features number: 600, sMAPE: 96.56433572921719
number of clusters: 2
sMAPE weighted average in clusters: 91.1974, entries: 1067

Features number: 900, sMAPE: 96.32938230438361
number of clusters: 2
sMAPE weighted average in clusters: 88.9605, entries: 1067
---------------------------------------

Searching parameters for label updrs_3

Features number: 300, sMAP

### note: after features reduction results with clustering significantly improved (84.17 vs. 90.52)

# Final results

In [128]:
def reduce_features_number(dataset, feat_nr):
    
    sorted_correlations = sort_label_correlations(dataset)
    sorted_features = list(sorted_correlations.keys())
    label = [dataset.columns[0]]

    cols = label + sorted_features[:feat_nr]
    df = dataset[cols]
    
    return df

In [129]:
def reduce_datasets_features(dataset1, dataset2, dataset3, dataset4, feat_nr_list):
    df1 = reduce_features_number(dataset1, feat_nr_list[0])
    df2 = reduce_features_number(dataset2, feat_nr_list[1])
    df3 = reduce_features_number(dataset3, feat_nr_list[2])
    df4 = reduce_features_number(dataset4, feat_nr_list[3])
    
    return df1, df2, df3, df4

### Searching parameters were performed for more variants, which resulted in the best set:
### * label updrs_1: 700 most correlated features and 2 clusters
### * label updrs_2: 700 most correlated features and 3 clusters
### * label updrs_3: 1100 most correlated features and 4 clusters
### * label updrs_4: 300 most correlated features and 2 clusters

In [138]:
df1_r, df2_r, df3_r, df4_r = reduce_datasets_features(df1, df2, df3, df4, [700, 700, 1100, 300])
subsets = get_training_subsets(df1_r, df2_r, df3_r, df4_r)
predict_labels(subsets)
predict_clustered_labels(df1_r, df2_r, df3_r, df4_r, [2, 3, 4, 2])


PREDICTION OF LABELS
sMAPE for label 0: 61.257784272052284
sMAPE for label 1: 97.808318433834
sMAPE for label 2: 83.40575127291633
sMAPE for label 3: 146.6678751951719

sMAPE weighted average of labels: 90.7649

PREDICTION OF CLUSTERED LABELS

LABEL: updrs_1
CLUSTER 0:sMAPE: 54.4426, entries: 663
CLUSTER 1:sMAPE: 55.5814, entries: 404
sMAPE weighted average in clusters: 54.8738, entries: 1067
------------------------------------------------

LABEL: updrs_2
CLUSTER 0:sMAPE: 86.4457, entries: 488
CLUSTER 1:sMAPE: 85.8088, entries: 426
CLUSTER 2:sMAPE: 82.5932, entries: 153
sMAPE weighted average in clusters: 85.6390, entries: 1067
------------------------------------------------

LABEL: updrs_3
CLUSTER 0:sMAPE: 79.6934, entries: 150
CLUSTER 1:sMAPE: 69.3884, entries: 282
CLUSTER 2:sMAPE: 79.7936, entries: 187
CLUSTER 3:sMAPE: 80.0440, entries: 438
sMAPE weighted average in clusters: 77.1071, entries: 1057
------------------------------------------------

LABEL: updrs_4
CLUSTER 0:sMAPE: 

# Conclusion: 
### * features reduction improved sMAPE without clustering (from 92.06 to 90.76)
### * features reduction with optimal clustering improved sMAPE to 81.67